monix-tail/shared/src/main/scala/monix/tail/internal/IterantSwitchIfEmpty.scala

```diff
@@ -0,0 +1,27 @@
+package monix.tail.internal
+
+import cats.syntax.functor._
+import cats.effect.Sync
+import monix.tail.Iterant
+import monix.tail.Iterant._
+
+private[tail] object IterantSwitchIfEmpty {
+  def apply[F[_], A](primary: Iterant[F, A], backup: Iterant[F, A])(implicit F: Sync[F]): Iterant[F, A] = {
+    def loop(source: Iterant[F, A]): Iterant[F, A] =
+      source match {
+        case Suspend(rest, stop) => Suspend(rest.map(loop), stop)
+        case NextBatch(batch, rest, stop) if batch.cursor().isEmpty =>
```

Comment:
Here we have a problem. As an implementation detail of `Iterant`, the `NextBatch` and `NextCursor` are wrapping the equivalent of an `Iterable` and `Iterator`, but once you start operating on iterators, you end up with side-effects.

So `batch.cursor().isEmpty` is side-effectful and we can't allow that without suspending it.
Take a look for example at [.take](https://github.com/monix/monix/blob/v3.0.0-M3/monix-tail/shared/src/main/scala/monix/tail/internal/IterantTake.scala#L88) where we are doing something like this:

```scala
    source match {
      case NextBatch(_, _, _) | NextCursor(_, _, _) =>
        // We can have side-effects with NextBatch/NextCursor
        // processing, so suspending execution in this case
        Suspend(F.delay(loop(n)(source)), source.earlyStop)
      case _ =>
        loop(n)(source)
    }
}
```

We need the same here.

Reply:
Another thing — once you evaluate `batch.cursor()`, given that it is side-effectful, it's wasteful to still return `source` in case the generated cursor is NOT empty.

It's better at this point to transform that node into a `NextCursor`. So...

```scala
case NextBatch(batch, rest, stop) =>
  val cursor = batch.cursor()
  if (cursor.isEmpty) 
     Suspend(rest.map(loop), stop)
  else
    NextCursor(cursor, rest, stop)
```

Reply:
And yet another implementation detail ... we assume that `batch.cursor()` is hostile and can throw (after all we expect side effects from it), the general idea being that you can take any dirty Java Iterable and wrap it into a batch or cursor.

So we also need a `try / catch` that wraps this pattern matching and that returns `Halt(error)`.
See [IterantTake](https://github.com/monix/monix/blob/v3.0.0-M3/monix-tail/shared/src/main/scala/monix/tail/internal/IterantTake.scala#L81) for this one as well.